src/compiler/checker.ts

```diff
@@ -9696,16 +9696,32 @@ namespace ts {
             }
         }
 
+        function isParameterOfContextSensitiveSignature(symbol: Symbol) {
+            let decl = symbol.valueDeclaration;
+            if (!decl) {
+                return false;
+            }
+            if (isBindingElement(decl)) {
+                decl = walkUpBindingElementsAndPatterns(decl);
+            }
+            if (isParameter(decl)) {
+                return (isExpression(decl.parent) || isObjectLiteralMethod(decl.parent)) && isContextSensitive(decl.parent);

```

Comment:
`isExpression` is going to do a lot more work than it needs to do. We have an `isContextSensitiveFunctionOrObjectLiteralMethod` which only checks for the 3 valid forms.

There's another caveat which is that for a JSDoc'd function declaration, [we consider that to be context-sensitive-ish](https://insiders.vscode.dev/github.com/microsoft/TypeScript/blob/63791f52d4e7a3bf461b974e94abd8cbb6b546c5/src/compiler/checker.ts#L17609). I don't know if that's another scenario we have to deal with.